### DISCUSSION TITLE
Appsilon rhino framework for shiny apps support in Positron IDE

### DISCUSSION THREAD

**Discussion by davidrsch** [posit-dev/positron#4822] [Feedback]
Hello. I have noticed while working with Positron IDE, that some functionalities, which are available in Rstudio IDE for building a shiny app using Appsilon's `rhino` framework, aren't available in Positron IDE. One of the most important aspects missing are addins.

> RStudio [Addins](https://rstudio.github.io/rstudioaddins/) provide a mechanism for executing R functions interactively from within the RStudio IDE either through keyboard shortcuts, or through the Addins menu.

![image](https://github.com/user-attachments/assets/5bc825e4-fefc-4632-8272-b2c42797597e)

I got used to work with them and don't having the possibility of using them in Positron IDE have being a little bit troublesome. There are other aspects like:
- Run shiny app isn't default for execution button in editor whe using rhino framework
   
![image](https://github.com/user-attachments/assets/faa3095b-d8e7-42f3-81c0-b8bc7b2d11a5)
Standard shiny app
![image](https://github.com/user-attachments/assets/325b11d1-21f2-47c3-93dc-8ac6192e7542)
Rhino app

- Running cypress tests. As can be seeing here https://github.com/Appsilon/rhino/issues/603#issue-2425728706
- `box.lsp` package for autocompletion when working with modularized R scripts using `box` package. (Add it here because `box` modules are how shiny modules are build using `rhino` framework)

I realize that some of this features being missing may be not entirely related to Positron IDE development. But with the existence of different shiny app development frameworks like `golem`, `leprechaun`, `teal`, `tapyr` (python) or even when using shiny-live in quarto the users must be capable of taking the most of them while using Positron. 

Note: Maybe it's a good idea take this to the shiny extension development.

**Comment by jmcphers** [comment]
> One of the most important aspects missing are addins.

We're tracking that work here: https://github.com/posit-dev/positron/issues/1313

Positron currently does not support Rhino or Cypress out of the box (and it is not likely we'd add native support in the near term). It would be possible for you or your team to build a Rhino extension for Positron, or request that the existing Shiny extension support it as you noted.

> box.lsp package for autocompletion 

The box.lsp package integrates with another R package called `languageserver`. This is the backend for the LSP in the VSCode-R plugin, but Positron doesn't use `languageserver`. Instead it has a new high-performance LSP for R embedded in our R kernel, ark. https://github.com/posit-dev/ark

Positron's infrastructure does allow multiple LSPs to run against the same editor surface, so if box.lsp could be refactored as a standalone LSP instead of an extension to `languageserver` it shouldn't be tough to get it to work in Positron.

**Comment by davidrsch** [comment]
Hello, I reviewed the shiny extension and found some possible solutions to the discussed.
- Addings: would be nice having an addition to the user interface to execute shiny extension snippets and allow other frameworks or eve user defined snippets https://github.com/posit-dev/shiny-vscode/issues/77#issue-2557358302.
- Run shiny app isn't default for execution button in editor when using rhino framework. This is because the extension looks for the keyword shiny in the app.R file, so I added a space in the comment of this file, changing it from:
   > # Rhino / shinyApp entrypoint. Do not edit.
   to:
   > # Rhino / shiny App entrypoint. Do not edit.
   and run button worked as expected.
   ![image](https://github.com/user-attachments/assets/96279d2c-fc53-4824-aa29-5b173d7112fb)
   Opened a discussion at rhino source for a similar change https://github.com/Appsilon/rhino/discussions/621#discussion-7260528. I would try to check if there are any other frameworks that have similar issue.
